mypy/semanal.py

```diff
@@ -3638,9 +3653,26 @@ def lookup_qualified(self, name: str, ctx: Context,
                     if n and n.module_hidden:
                         self.name_not_defined(name, ctx)
             if n and not n.module_hidden:
+                n = self.rebind_symbol_table_node(n)
                 return n
             return None
 
+    def rebind_symbol_table_node(self, n: SymbolTableNode) -> Optional[SymbolTableNode]:
+        """If node refers to old version of module, return reference to new version.
+
+        If the reference is removed in the new version, return None.
+        """
+        # TODO: Handle type aliases, type variables and other sorts of references
+        if isinstance(n.node, (FuncDef, OverloadedFuncDef, TypeInfo, Var)):
+            # TODO: Why is it possible for fullname() to be None, even though it's not
+            #   annotated as Optional[str]?
+            if n.node.fullname() and get_prefix(n.node.fullname()) == self.cur_mod_id:
```

Comment:
Does this need the same TODO as above on line 3642? ("Do this for all modules in the set of modified files.")

Also what if it's inside a class or other namespace? get_prefix() only strips one component from the end. Maybe we need a utility to compare a name with a prefix (basically `name.startswith(prefix + '.')`).